src/coins.cpp

```diff
@@ -27,14 +27,22 @@ bool CCoinsViewBacked::HaveCoin(const COutPoint &outpoint) const { return base->
 uint256 CCoinsViewBacked::GetBestBlock() const { return base->GetBestBlock(); }
 std::vector<uint256> CCoinsViewBacked::GetHeadBlocks() const { return base->GetHeadBlocks(); }
 void CCoinsViewBacked::SetBackend(CCoinsView &viewIn) { base = &viewIn; }
-bool CCoinsViewBacked::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock, bool erase) { return base->BatchWrite(mapCoins, hashBlock, erase); }
+bool CCoinsViewBacked::BatchWrite(CoinsViewCacheCursor& cursor, const uint256 &hashBlock) { return base->BatchWrite(cursor, hashBlock); }
 std::unique_ptr<CCoinsViewCursor> CCoinsViewBacked::Cursor() const { return base->Cursor(); }
 size_t CCoinsViewBacked::EstimateSize() const { return base->EstimateSize(); }
 
 CCoinsViewCache::CCoinsViewCache(CCoinsView* baseIn, bool deterministic) :
     CCoinsViewBacked(baseIn), m_deterministic(deterministic),
     cacheCoins(0, SaltedOutpointHasher(/*deterministic=*/deterministic), CCoinsMap::key_equal{}, &m_cache_coins_memory_resource)
-{}
+{
+    m_sentinel.second.SelfRef(m_sentinel);
+}
+
+CCoinsViewCache::~CCoinsViewCache()
+{
+    cacheCoins.clear();
+    Assume(m_sentinel.second.Next() == &m_sentinel);
```

Comment:
Is this destructor defined just to document this assumption?

Reply:
It is actually here to clear `cacheCoins` and make sure all `CCoinsCacheEntry`s are destroyed and call `ClearFlags` in their destructors. Since the first and last flagged entries have a reference to `m_sentinel`, we must be sure they remove themselves from the linked list and set `m_flags` to `0` before `m_sentinel` is destroyed.
Without this destructor, we can make this requirement implicit by having `m_sentinel` come before `coinsCache` in the `CCoinsViewCache` members. That way `m_sentinel` will be destroyed after `coinsCache` and all its entries have already been destroyed. But, that is brittle since someone could reorder the fields. By having this destructor here we make sure all entries are unflagged before destroying the sentinel.

Reply:
This was stumping me for a while, thank you!!

Reply:
Perhaps I should comment that here in a follow-up then?

Reply:
Follow-up: Why wouldn't `m_sentinel` being destroyed first be OK since: when `m_sentinel.second.ClearFlags()` happens, that makes the last (youngest) entry's `m_next` point to the first (oldest) entry and vice versa?

Reply:
`m_sentinel` does not have `m_flags` set. So when `m_sentinel` is destroyed it doesn't actually remove itself from the list.

Reply:
Thanks, I missed that, I think a comment in a follow-up would be helpful.

Reply:
Added a comment

Reply:
Now that we mark the sentinel as dirty, my response to your follow-up is outdated. So perhaps we don't need this destructor anymore? The last entry to be destroyed will just point to itself in `ClearFlags`, right?

Reply:
I removed the CCoinsViewCache destructor in light of this. Thank you!